### ISSUE TITLE
Assert Array Index

### ISSUE THREAD

**Issue by uid11** [type-challenges/type-challenges#925]
## Info

Basic info of your challenge questions,

```yaml
difficulty: extreme
title: Assert Array Index
tags: array
tsconfig: { noUncheckedIndexedAccess: true }
```

## Question

Sometimes we want to use the good old `for`-loop with an index to traverse the array, but in this case TypeScript does not check in any way that we are accessing the elements of the array at its real index (not exceeding the length of the array), and that we are not using an arbitrary number as an index, or index from another array (for nested loops, for traversing matrices or graphs):
```ts
const matrix = [
    [3, 4],
    [5, 6],
    [7, 8],
];

// This example contains no type errors when the noUncheckedIndexedAccess option is off.
for (let i = 0; i < matrix.length; i += 1) {
    const columns: number[] = matrix[i];

    for (let j = 0; j < columns.length; j += 1) {
        const current: number = columns[i]; // oops! i instead of j

        console.log(
            current.toFixed(), // TypeError: Cannot read property 'toFixed' of undefined
        );
    }
}
```

You can enable the [noUncheckedIndexedAccess](https://www.typescriptlang.org/tsconfig#noUncheckedIndexedAccess) option (in `tsconfig.json`), but then each time you access an array element, you will need to check that this element exists, which is somewhat verbose and inconvenient, especially since in the case of such a `for`-traversal, we are sure that the index does not exceed the length of the array:
```ts
const numbers = [5, 7];

for (let i = 0; i < numbers.length; i += 1) {
    const current = numbers[i];

    if (current !== undefined) {
        console.log(current.toFixed());
    }
}
```

Write an `assert`-function `assertArrayIndex(array, key)` that can be applied to any `array` (with an arbitrary unique string `key`, which is needed to distinguish arrays at the type level) to allow access to the elements of this array only by the index obtained from array by the special generic type `Index<typeof array>` (this functionality requires enabling the [noUncheckedIndexedAccess](https://www.typescriptlang.org/tsconfig#noUncheckedIndexedAccess) option in `tsconfig.json`):
```ts
const numbers = [5, 7];

assertArrayIndex(numbers, 'numbers');

for (let i = 0 as Index<typeof numbers>; i < numbers.length; i += 1) {
    console.log(numbers[i].toFixed());
}
```

When accessing by such an index, it must be guaranteed that an element in the array exists, and when accessing an array by any other indices, there is no such guarantee (the element may not exist):
```ts
const matrix = [
    [3, 4],
    [5, 6],
    [7, 8],
];

assertArrayIndex(matrix, 'rows');

let sum = 0;

for (let i = 0 as Index<typeof matrix>; i < matrix.length; i += 1) {
    const columns: number[] = matrix[i];

    // @ts-expect-error: number | undefined in not assignable to number
    const x: number[] = matrix[0];

    assertArrayIndex(columns, 'columns');

    for (let j = 0 as Index<typeof columns>; j < columns.length; j += 1) {
        sum += columns[j];

        // @ts-expect-error: number | undefined in not assignable to number
        const y: number = columns[i];

        // @ts-expect-error: number | undefined in not assignable to number
        const z: number = columns[0];

        // @ts-expect-error: number[] | undefined in not assignable to number[]
        const u: number[] = matrix[j];
    }
}
```

The `assertArrayIndex` function cannot be called on tuples (since the accessing the elements is already well typed in them):
```ts
const tuple = [5, 7] as const;

// @ts-expect-error
assertArrayIndex(tuple, 'tuple');
```

## Template

This is the template for challengers to start the coding. Basically, you just need to change the name of your generic/function and leave to implementation `any`.

```ts
function assertArrayIndex(array: readonly unknown[], key: string) {}

type Index<Array> = any
```

## Test Cases

Provide some test cases for your challenge, you can use some utils from `@type-challenges/utils` for asserting.  

```ts
const matrix = [
    [3, 4],
    [5, 6],
    [7, 8],
];

assertArrayIndex(matrix, 'rows');

let sum = 0;

for (let i = 0 as Index<typeof matrix>; i < matrix.length; i += 1) {
    const columns: number[] = matrix[i];

    // @ts-expect-error: number | undefined in not assignable to number
    const x: number[] = matrix[0];

    assertArrayIndex(columns, 'columns');

    for (let j = 0 as Index<typeof columns>; j < columns.length; j += 1) {
        sum += columns[j];

        // @ts-expect-error: number | undefined in not assignable to number
        const y: number = columns[i];

        // @ts-expect-error: number | undefined in not assignable to number
        const z: number = columns[0];

        // @ts-expect-error: number[] | undefined in not assignable to number[]
        const u: number[] = matrix[j];
    }
}

const a: string[] = [];

assertArrayIndex(a, 'a');

for (let p = 0 as Index<typeof a>; p < a.length; p += 1) {
    const value: string = a[p];

    // @ts-expect-error: string | undefined is not assignable to string
    const z: string = a[2];
}

a.push('qux');
// @ts-expect-error: number is not assignable to string
a.push(3);

for (const value of a) {
    const other: string = value;
}

const b: number[] = [];

assertArrayIndex(b, 'b');

for (let p = 0 as Index<typeof a>; p < b.length; p += 1) {
    // @ts-expect-error: number | undefined is not assignable to string
    const value: string = b[p];
}

const c: string[] = [];

assertArrayIndex(c, 'c');

for (let p = 0; p < c.length; p += 1) {
    // @ts-expect-error: string | undefined is not assignable to string
    let value: string = c[p];

    // @ts-expect-error: string | undefined is not assignable to string
    value = c[0 as Index<typeof a>];
}

const d: readonly number[] = [];

assertArrayIndex(d, 'd');

for (let p = 0 as Index<typeof d>; p < d.length; p += 1) {
    const value: number = d[p];

    // @ts-expect-error: only permits reading
    d[2] = 3;
}

// @ts-expect-error: push does not exist on readonly
d.push(3);

const e: [number] = [0];

// @ts-expect-error: [number] is not assignable to never
assertArrayIndex(e, 'e');

const f: readonly [boolean] = [false];

// @ts-expect-error: [boolean] is not assignable to never
assertArrayIndex(f, 'f');

const tuple = [5, 7] as const;

// @ts-expect-error: readonly [5, 7] is not assignable to never
assertArrayIndex(tuple, 'tuple');
```

**Comment by antfu** [comment]
I am not very sure about this, does this requires runtime code? Also, I am not sure the solution to this problem is ideal, maybe we can find a better way?

**Comment by uid11** [comment]
I understand your concerns, since the proposed approach looks really unusual. I will try to answer the points.

First, what problem are we solving (I note right away that this is not a very important or acute problem)?
In TypeScript there is a problem with access to the elements of the array `T[]` -- without `noUncheckedIndexedAccess` the compiler returns `T` at any index, and with `noUncheckedIndexedAccess` it always returns `T | undefined`. Moreover, in both cases, the compiler does not check in any way that the index does not exceed the size of the array.

That is, the real problem is that having an array, we can read from it at any index (any `number`), and the array itself will not warn us about an error in any way.

A natural solution that has long been used in such cases in TypeScript is to use branded types. For each required array, we can create some kind of its branded index, and somehow transform (wrap) the array, forcing it to accept only this branded index.

The next question is only in an ergonomic and convenient API for creating such arrays and their indices.

The first decision that comes to mind is to make a generic that will return a certain subtype of the array (a subtype so that you can use array methods on it, and also convert real arrays to this type), and a separate generic that will get the index type for this subtype:
```ts
type MyArray = IndexedArray<number, 'MyArray'> // number[] with branded index
type MyIndex = Index<MyArray> // branded index itself
```

This is a working solution, but it has its own difficulties -- arrays usually come to us from external sources (and we, of course, do not want to distribute a private branded subtype of an array across the API, among other modules), and therefore, before starting the loop, we have to cast arrays to the desired subtype (also using `typeof` to extract the type of the array):
```ts
const myArray: IndexedArray<typeof externalArray[number], 'MyArray'> = externalArray
let index: Index<typeof myArray> = 0
```

I tried such approaches, but it seemed to me more convenient to use the built-in TypeScript mechanism of [assert](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions)-functions (this is an important mechanism, but so far there have been no challenges about it, if I'm not mistaken). This allows us to narrow the type of an existing array (no duplication and `typeof`), and furthermore, such a narrowing only works in the corresponding branch in the code flow, hiding the branded subtype information from the rest of the project where this subtype is superfluous:
```ts
assertArrayIndex(externalArray, 'MyArray')
let index: Index<typeof externalArray> = 0
```

The main disadvantage of `assert`-functions is their impact on runtime, but in reality this is an imaginary disadvantage.

An `assert`-function that does not perform any runtime checks (this is our case) turns into a function with an empty body after compilation, and modern JavaScript engines perfectly optimize calls to such functions (in fact, cut them out) -- especially since this function will be called once for each traversal array, that is, not often at all.
```js
// this is what the assert-function looks like in the emitted code
export function assertArrayIndex(array, key) {}
```

Finally, modern code bundlers can strip such functions statically (for example, in `webpack` you can do this with the `compress.pure_funcs` option for [terser](https://github.com/terser/terser#command-line-options), which is built into the `webpack` as the default code [minimizer](https://webpack.js.org/configuration/optimization/#optimizationminimizer)).

In other words, I tried different approaches to solving the described problem, and it was the `assert`-function approach that seemed to me the most convenient.

I wanted to offer it as a challenge to show how the idea of ​​branded primitive types can evolve further, so that users know they can use similar solutions if needed.

**Comment by zhaoyao91** [comment]
so why not return `Index<typeof externalArray>` by a company function with `assertArrayIndex` such as `arrayIndex(0)`?
